### PR TITLE
feat: support subscription periods (yearly/monthly)

### PR DESCRIPTION
**Pull request by Chinoholo0807**

### CODE REVIEW

**Comment by Chinoholo0807** [comment]
https://github.com/casdoor/casdoor/assets/54203997/961e0cf2-6ed5-44a8-a225-8203adad2708

File: web/src/pricing/PricingPage.js [hunk 1]
```diff
@@ -161,18 +212,25 @@ class PricingPage extends React.Component {
...
+              </Row>
               <Row style={{width: "100%", marginTop: "40px"}}>
                 <Col span={24} style={{display: "flex", justifyContent: "center"}} >
                   {
                     this.renderCards()
                   }
                 </Col>
               </Row>
-              <Row style={{justifyContent: "center"}}>
+              {/* <Row style={{justifyContent: "center"}}>
```
Comment (hsluoyz) [inline web/src/pricing/PricingPage.js:229]: Don't remove trial

File: web/src/pricing/SingleCard.js [hunk 2]
```diff
@@ -39,20 +38,23 @@ class SingleCard extends React.Component {
...
-                <span style={{fontSize: "40px", fontWeight: 700}}>{Setting.getCurrencySymbol(plan.currency)} {plan.pricePerYear}</span>
-                <span style={{fontSize: "18px", fontWeight: 600, color: "gray"}}>  {i18next.t("plan:per year")}</span>
-              </div>
-            </Row>
-
+            {
+              mode === "month" ? (
+                <Row>
+                  <div style={{textAlign: "left"}} className="px-10 mt-5">
+                    <span style={{fontSize: "40px", fontWeight: 700}}>{Setting.getCurrencySymbol(plan.currency)} {plan.pricePerMonth}</span>
```
Comment (hsluoyz) [inline web/src/pricing/SingleCard.js:45]: `{mode === "month" ? plan.pricePerMonth : plan.pricePerYear}`

File: web/src/pricing/PricingPage.js [hunk 3]
```diff
@@ -83,29 +85,71 @@ class PricingPage extends React.Component {
       });
   }
 
-  loadPricing(pricingName) {
-    if (pricingName === undefined) {
+  async loadPricing(pricingName) {
```
Comment (hsluoyz) [inline web/src/pricing/PricingPage.js:88]: Why add async?

File: object/pricing.go [hunk 4]
```diff
@@ -29,15 +29,10 @@ type Pricing struct {
 	Description string `xorm:"varchar(100)" json:"description"`
 
 	Plans         []string `xorm:"mediumtext" json:"plans"`
+	Modes         []string `json:"modes"`
```
Comment (hsluoyz) [inline object/pricing.go:32]: `Mode string`

Three options:

“Monthly”, "Yearly", "Monthly and Yearly"

File: web/src/PricingEditPage.js [hunk 5]
```diff
@@ -195,6 +195,26 @@ class PricingEditPage extends React.Component {
           </Col>
         </Row>
         <Row style={{marginTop: "20px"}} >
+          <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+            {Setting.getLabel(i18next.t("general:Modes"), i18next.t("general:Modes - Tooltip"))} :
+          </Col>
+          <Col span={22} >
+            <Checkbox.Group
+              options={[{label: "Monthly", value: "month"}, {label: "Yearly", value: "year"}]}
```
Comment (hsluoyz) [inline web/src/PricingEditPage.js:203]: {label: "Monthly", value: "Monthly"}, {label: "Yearly", value: "Yearly"}

File: object/plan.go [hunk 6]
```diff
@@ -31,7 +33,8 @@ type Plan struct {
 	PricePerMonth    float64  `json:"pricePerMonth"`
 	PricePerYear     float64  `json:"pricePerYear"`
 	Currency         string   `xorm:"varchar(100)" json:"currency"`
-	Product          string   `json:"product"`                              // related product id
+	ProductMonth     string   `xorm:"'product'" json:"productMonth"`        // related product id (sub mode is month)
```
Comment (hsluoyz) [inline object/plan.go:36]: ProductMonthly, remove comment

File: web/src/pricing/PricingPage.js (hunk 3 above)
Comment (Chinoholo0807) [reply web/src/pricing/PricingPage.js:89]: use await in function

**Comment by Chinoholo0807** [comment]
https://github.com/casdoor/casdoor/assets/54203997/1be05f53-b387-4791-999b-c39149a1a015

File: object/pricing.go [hunk 7]
```diff
@@ -21,23 +21,25 @@ import (
 	"github.com/xorm-io/core"
 )
 
+type Period string
```
Comment (hsluoyz) [inline object/pricing.go:24]: Just string

File: web/src/pricing/SingleCard.js [hunk 8]
```diff
@@ -37,43 +37,59 @@ class SingleCard extends React.Component {
...
         >
-          <div style={{textAlign: "left"}} className="px-10 mt-5">
-            <span style={{fontSize: "40px", fontWeight: 700}}>{Setting.getCurrencySymbol(plan.currency)} {plan.pricePerMonth}</span>
-            <span style={{fontSize: "18px", fontWeight: 600, color: "gray"}}>  {i18next.t("plan:per month")}</span>
-          </div>
+          <Col>
+            <Row>
+              <div style={{textAlign: "left"}} className="px-10 mt-5">
+                {
+                  period === "Monthly" ? (
```
Comment (hsluoyz) [inline web/src/pricing/SingleCard.js:44]: Can't you write this way?

```react
<>
  <span style={{fontSize: "40px", fontWeight: 700}}>{Setting.getCurrencySymbol(plan.currency)} {period === "Monthly" ? plan.pricePerMonth : plan.pricePerYearly}</span>
    <span style={{fontSize: "18px", fontWeight: 600, color: "gray"}}>  {i18next.t("plan:per month")}</span>
  </>
```

File: web/src/pricing/PricingPage.js [hunk 9]
```diff
@@ -136,7 +167,14 @@ class PricingPage extends React.Component {
             {
               this.state.plans.map(item => {
                 return (
-                  <SingleCard style={{marginRight: "5px", marginLeft: "5px"}} link={getUrlByPlan(item.name)} key={item.name} plan={item} isSingle={this.state.plans.length === 1} />
+                  <SingleCard
```
Comment (hsluoyz) [inline web/src/pricing/PricingPage.js:170]: Don't change unrelated things

File: object/plan.go [hunk 10]
```diff
@@ -31,7 +32,8 @@ type Plan struct {
 	PricePerMonth    float64  `json:"pricePerMonth"`
 	PricePerYear     float64  `json:"pricePerYear"`
 	Currency         string   `xorm:"varchar(100)" json:"currency"`
-	Product          string   `json:"product"`                              // related product id
+	ProductMonth     string   `xorm:"'product'" json:"productMonth"`
```
Comment (hsluoyz) [inline object/plan.go:35]: Remove ProductMonth and ProductYear, add `Period` as string

File: object/pricing.go [hunk 11]
```diff
@@ -21,23 +21,25 @@ import (
...
-	TrialDuration int      `json:"trialDuration"`
-	Application   string   `xorm:"varchar(100)" json:"application"`
-
-	Submitter   string `xorm:"varchar(100)" json:"submitter"`
-	Approver    string `xorm:"varchar(100)" json:"approver"`
-	ApproveTime string `xorm:"varchar(100)" json:"approveTime"`
-
-	State string `xorm:"varchar(100)" json:"state"`
+	Plans          []string `xorm:"mediumtext" json:"plans"`
+	BillingPeriods []Period `json:"billingPeriods"`
```
Comment (hsluoyz) [inline object/pricing.go:39]: Remove `BillingPeriods`

File: object/product.go [hunk 12]
```diff
@@ -158,7 +158,7 @@ func (product *Product) getProvider(providerName string) (*Provider, error) {
 	return provider, nil
 }
 
-func BuyProduct(id string, user *User, providerName, pricingName, planName, host string) (string, string, error) {
+func BuyProduct(id string, user *User, providerName, pricingName, planName, host string, period Period) (string, string, error) {
```
Comment (hsluoyz) [inline object/product.go:161]: Don't add `period` as arg everywhere, it's not that important

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/subscription.go [hunk 13]
```diff
@@ -116,9 +118,10 @@ func NewSubscription(owner, userName, pricingName, planName, paymentName string)
 		Plan:    planName,
```
Comment (hsluoyz) [inline object/subscription.go:118]: Subscription doesn't need to know pricing?

File: object/subscription.go [hunk 14]
```diff
@@ -116,9 +118,10 @@ func NewSubscription(owner, userName, pricingName, planName, paymentName string)
...
 		Payment: paymentName,
 
-		StartTime: time.Now(),
-		EndTime:   time.Now().AddDate(0, 0, 30),
-		Duration:  30,              // TODO
+		StartTime: startTime,
+		EndTime:   endTime,
+		Duration:  duration,
```
Comment (hsluoyz) [inline object/subscription.go:123]: Duration and Period are duplicated, keep `Period`

File: web/src/PricingEditPage.js [hunk 15]
```diff
@@ -194,6 +194,22 @@ class PricingEditPage extends React.Component {
             />
           </Col>
         </Row>
+        <Row style={{marginTop: "20px"}} >
+          <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
```
Comment (hsluoyz) [inline web/src/PricingEditPage.js:198]: Remove

File: web/src/Setting.js [hunk 16]
```diff
@@ -1264,3 +1264,10 @@ export function getCurrencySymbol(currency) {
     return currency;
   }
 }
+
+export function getValidArray(array, defaultValue) {
```
Comment (hsluoyz) [inline web/src/Setting.js:1268]: Remove

File: web/src/auth/SignupPage.js [hunk 17]
```diff
@@ -136,7 +136,7 @@ class SignupPage extends React.Component {
   getResultPath(application, signupParams) {
     if (signupParams?.plan && signupParams?.pricing) {
       // the prompt page needs the user to be signed in, so for paid-user sign up, just go to buy-plan page
-      return `/buy-plan/${application.organization}/${signupParams?.pricing}?user=${signupParams.username}&plan=${signupParams.plan}`;
+      return `/buy-plan/${application.organization}/${signupParams?.pricing}?user=${signupParams.username}&plan=${signupParams.plan}&period=${signupParams.period}`;
```
Comment (hsluoyz) [inline web/src/auth/SignupPage.js:139]: Remove

File: web/src/pricing/PricingPage.js [hunk 18]
```diff
@@ -83,31 +86,59 @@ class PricingPage extends React.Component {
...
-          Setting.showMessage("error", res.msg);
-          return;
-        }
-        this.setState({
-          loading: false,
-          pricing: res.data,
-        });
-        this.onUpdatePricing(res.data);
+    try {
+      const res = await PricingBackend.getPricing(this.state.owner, pricingName);
```
Comment (hsluoyz) [inline web/src/pricing/PricingPage.js:94]: Why not use callback?

File: web/src/pricing/PricingPage.js [hunk 19]
```diff
@@ -83,31 +86,59 @@ class PricingPage extends React.Component {
...
+      if (res.status === "error") {
+        throw new Error(res.msg);
```
Comment (hsluoyz) [inline web/src/pricing/PricingPage.js:96]: Why change to throw? Throw to console?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@Chinoholo0807 fix:

image

**Comment by Chinoholo0807** [comment]
Uploading 2023-08-30_15-06-02.mp4…

**Pull request merged**